File: src/mobile/lib/features/battles/models/battle_model.dart
/// 1v1 Battle Model
class Battle1v1 {
  final String id;
  final String matchCode;
  final String topicId;
  final String categoryId;
  final String challengerId;
  final String? opponentId;
  final int questionCount;
  final int timePerQuestion;
  final int entryCoins;
  final int prizeCoins;
  final List<BattleQuestion>? questions;
  final String status;
  final String? winnerId;
  final bool isDraw;
  final DateTime? expiresAt;
  final DateTime? startedAt;
  final DateTime? endedAt;

  // Challenger stats
  final int challengerScore;
  final int challengerCorrect;
  final int challengerTimeMs;
  final bool challengerReady;

  // Opponent stats
  final int opponentScore;
  final int opponentCorrect;
  final int opponentTimeMs;
  final bool opponentReady;

  // Player info
  final BattlePlayer? challenger;
  final BattlePlayer? opponent;

  Battle1v1({
    required this.id,
    required this.matchCode,
    required this.topicId,
    required this.categoryId,
    required this.challengerId,
    this.opponentId,
    this.questionCount = 10,
    this.timePerQuestion = 15,
    this.entryCoins = 0,
    this.prizeCoins = 0,
    this.questions,
    this.status = 'waiting',
    this.winnerId,
    this.isDraw = false,
    this.expiresAt,
    this.startedAt,
    this.endedAt,
    this.challengerScore = 0,
    this.challengerCorrect = 0,
    this.challengerTimeMs = 0,
    this.challengerReady = false,
    this.opponentScore = 0,
    this.opponentCorrect = 0,
    this.opponentTimeMs = 0,
    this.opponentReady = false,
    this.challenger,
    this.opponent,
  });

  factory Battle1v1.fromJson(Map<String, dynamic> json) {
    List<BattleQuestion>? questions;
    if (json['questions'] != null && json['questions'] is List) {
      questions = (json['questions'] as List<dynamic>)
          .map((q) => BattleQuestion.fromJson(q as Map<String, dynamic>))
          .toList();
    }

    return Battle1v1(
      id: json['id']?.toString() ?? '',
      matchCode: (json['match_code'] as String?) ?? '',
      topicId: json['topic_id']?.toString() ?? '',
      categoryId: json['category_id']?.toString() ?? '',
      challengerId: json['challenger_id']?.toString() ?? '',
      opponentId: json['opponent_id']?.toString(),
      questionCount: int.tryParse(json['question_count']?.toString() ?? '10') ?? 10,
      timePerQuestion: int.tryParse(json['time_per_question']?.toString() ?? '15') ?? 15,
      entryCoins: int.tryParse(json['entry_coins']?.toString() ?? '0') ?? 0,
      prizeCoins: int.tryParse(json['prize_coins']?.toString() ?? '0') ?? 0,
      questions: questions,
      status: (json['status'] as String?) ?? 'waiting',
      winnerId: json['winner_id']?.toString(),
      isDraw: json['is_draw'] == '1' || json['is_draw'] == true,
      expiresAt: json['expires_at'] != null ? DateTime.tryParse(json['expires_at'] as String) : null,
      startedAt: json['started_at'] != null ? DateTime.tryParse(json['started_at'] as String) : null,
      endedAt: json['ended_at'] != null ? DateTime.tryParse(json['ended_at'] as String) : null,
      challengerScore: int.tryParse(json['challenger_score']?.toString() ?? '0') ?? 0,
      challengerCorrect: int.tryParse(json['challenger_correct']?.toString() ?? '0') ?? 0,
      challengerTimeMs: int.tryParse(json['challenger_time_ms']?.toString() ?? '0') ?? 0,
      challengerReady: json['challenger_ready'] == '1' || json['challenger_ready'] == true,
      opponentScore: int.tryParse(json['opponent_score']?.toString() ?? '0') ?? 0,
      opponentCorrect: int.tryParse(json['opponent_correct']?.toString() ?? '0') ?? 0,
      opponentTimeMs: int.tryParse(json['opponent_time_ms']?.toString() ?? '0') ?? 0,
      opponentReady: json['opponent_ready'] == '1' || json['opponent_ready'] == true,
      challenger: json['challenger'] != null ? BattlePlayer.fromJson(json['challenger'] as Map<String, dynamic>) : null,
      opponent: json['opponent'] != null ? BattlePlayer.fromJson(json['opponent'] as Map<String, dynamic>) : null,
    );
  }

  bool get isWaiting => status == 'waiting';
  bool get isReady => status == 'ready';
  bool get isPlaying => status == 'playing';
  bool get isCompleted => status == 'completed';
  bool get isExpired => status == 'expired';

  bool isChallenger(String oderId) => challengerId == oderId;
  bool isOpponent(String oderId) => opponentId == oderId;
  bool isWinner(String oderId) => winnerId == oderId;
}

/// Battle Player Info
class BattlePlayer {
  final String id;
  final String? name;
  final String? profile;

  BattlePlayer({
    required this.id,
    this.name,
    this.profile,
  });

  factory BattlePlayer.fromJson(Map<String, dynamic> json) {
    return BattlePlayer(
      id: json['id']?.toString() ?? '',
      name: json['name'] as String?,
      profile: json['profile'] as String?,
    );
  }
}

/// Battle Question
class BattleQuestion {
  final String id;
  final String question;
  final String optionA;
  final String optionB;
  final String? optionC;
  final String? optionD;
  final String answer;
  final String? note;

  BattleQuestion({
    required this.id,
    required this.question,
    required this.optionA,
    required this.optionB,
    this.optionC,
    this.optionD,
    required this.answer,
    this.note,
  });

  factory BattleQuestion.fromJson(Map<String, dynamic> json) {
    return BattleQuestion(
      id: json['id']?.toString() ?? '',
      question: (json['question'] as String?) ?? '',
      optionA: (json['optiona'] as String?) ?? '',
      optionB: (json['optionb'] as String?) ?? '',
      optionC: json['optionc'] as String?,
      optionD: json['optiond'] as String?,
      answer: (json['answer'] as String?) ?? '',
      note: json['note'] as String?,
    );
  }

  List<String> get options {
    final opts = [optionA, optionB];
    if (optionC != null && optionC!.isNotEmpty) opts.add(optionC!);
    if (optionD != null && optionD!.isNotEmpty) opts.add(optionD!);
    return opts;
  }
}

/// Group Battle Model
class GroupBattle {
  final String id;
  final String groupId;
  final String topicId;
  final String categoryId;
  final String createdBy;
  final String? title;
  final int questionCount;
  final int timePerQuestion;
  final int entryCoins;
  final int prizeCoins;
  final int minPlayers;
  final int maxPlayers;
  final int playerCount;
  final String status;
  final List<BattleQuestion>? questions;
  final List<GroupBattleEntry>? entries;
  final DateTime? scheduledStart;
  final DateTime? startedAt;
  final DateTime? endedAt;

  GroupBattle({
    required this.id,
    required this.groupId,
    required this.topicId,
    required this.categoryId,
    required this.createdBy,
    this.title,
    this.questionCount = 10,
    this.timePerQuestion = 15,
    this.entryCoins = 0,
    this.prizeCoins = 0,
    this.minPlayers = 2,
    this.maxPlayers = 10,
    this.playerCount = 0,
    this.status = 'pending',
    this.questions,
    this.entries,
    this.scheduledStart,
    this.startedAt,
    this.endedAt,
  });

  factory GroupBattle.fromJson(Map<String, dynamic> json) {
    List<BattleQuestion>? questions;
    if (json['questions'] != null && json['questions'] is List) {
      questions = (json['questions'] as List<dynamic>)
          .map((q) => BattleQuestion.fromJson(q as Map<String, dynamic>))
          .toList();
    }

    List<GroupBattleEntry>? entries;
    if (json['entries'] != null && json['entries'] is List) {
      entries = (json['entries'] as List<dynamic>)
          .map((e) => GroupBattleEntry.fromJson(e as Map<String, dynamic>))
          .toList();
    }

    return GroupBattle(
      id: json['id']?.toString() ?? '',
      groupId: json['group_id']?.toString() ?? '',
      topicId: json['topic_id']?.toString() ?? '',
      categoryId: json['category_id']?.toString() ?? '',
      createdBy: json['created_by']?.toString() ?? '',
      title: json['title'] as String?,
      questionCount: int.tryParse(json['question_count']?.toString() ?? '10') ?? 10,
      timePerQuestion: int.tryParse(json['time_per_question']?.toString() ?? '15') ?? 15,
      entryCoins: int.tryParse(json['entry_coins']?.toString() ?? '0') ?? 0,
      prizeCoins: int.tryParse(json['prize_coins']?.toString() ?? '0') ?? 0,
      minPlayers: int.tryParse(json['min_players']?.toString() ?? '2') ?? 2,
      maxPlayers: int.tryParse(json['max_players']?.toString() ?? '10') ?? 10,
      playerCount: int.tryParse(json['player_count']?.toString() ?? '0') ?? 0,
      status: (json['status'] as String?) ?? 'pending',
      questions: questions,
      entries: entries,
      scheduledStart: json['scheduled_start'] != null
          ? DateTime.tryParse(json['scheduled_start'] as String)
          : null,
      startedAt: json['started_at'] != null ? DateTime.tryParse(json['started_at'] as String) : null,
      endedAt: json['ended_at'] != null ? DateTime.tryParse(json['ended_at'] as String) : null,
    );
  }

  bool get isPending => status == 'pending';
  bool get isActive => status == 'active';
  bool get isCompleted => status == 'completed';
  bool get canStart => playerCount >= minPlayers;
}

/// Group Battle Entry
class GroupBattleEntry {
  final String id;
  final String battleId;
  final String oderId;
  final int score;
  final int correctAnswers;
  final int wrongAnswers;
  final int totalTimeMs;
  final int? rank;
  final int coinsEarned;
  final String status;
  final String? name;
  final String? profile;

  GroupBattleEntry({
    required this.id,
    required this.battleId,
    required this.oderId,
    this.score = 0,
    this.correctAnswers = 0,
    this.wrongAnswers = 0,
    this.totalTimeMs = 0,
    this.rank,
    this.coinsEarned = 0,
    this.status = 'joined',
    this.name,
    this.profile,
  });

  factory GroupBattleEntry.fromJson(Map<String, dynamic> json) {
    return GroupBattleEntry(
      id: json['id']?.toString() ?? '',
      battleId: json['battle_id']?.toString() ?? '',
      oderId: json['user_id']?.toString() ?? '',
      score: int.tryParse(json['score']?.toString() ?? '0') ?? 0,
      correctAnswers: int.tryParse(json['correct_answers']?.toString() ?? '0') ?? 0,
      wrongAnswers: int.tryParse(json['wrong_answers']?.toString() ?? '0') ?? 0,
      totalTimeMs: int.tryParse(json['total_time_ms']?.toString() ?? '0') ?? 0,
      rank: int.tryParse(json['rank']?.toString() ?? ''),
      coinsEarned: int.tryParse(json['coins_earned']?.toString() ?? '0') ?? 0,
      status: (json['status'] as String?) ?? 'joined',
      name: json['name'] as String?,
      profile: json['profile'] as String?,
    );
  }

  bool get isWinner => rank == 1;
}

File: src/admin/public/application/models/Battle_model.php
<?php

defined('BASEPATH') or exit('No direct script access allowed');

/**
 * Battle Model
 * Handles 1v1 and group battle operations
 */
class Battle_model extends CI_Model
{
    // ============================================
    // 1v1 BATTLES
    // ============================================

    /**
     * Create a 1v1 battle
     */
    public function create_1v1_battle($challenger_id, $topic_id, $category_id, $config = [])
    {
        $match_code = $this->generate_match_code();
        
        $data = [
            'match_code' => $match_code,
            'topic_id' => $topic_id,
            'category_id' => $category_id,
            'challenger_id' => $challenger_id,
            'question_count' => $config['question_count'] ?? 10,
            'time_per_question' => $config['time_per_question'] ?? 15,
            'entry_coins' => $config['entry_coins'] ?? 0,
            'prize_coins' => $config['prize_coins'] ?? 0,
            'status' => 'waiting',
            'expires_at' => date('Y-m-d H:i:s', strtotime('+10 minutes'))
        ];
        
        // Get random questions
        $questions = $this->get_random_questions($category_id, $data['question_count']);
        $data['questions'] = json_encode($questions);
        
        $this->db->insert('tbl_battle_1v1', $data);
        
        return [
            'battle_id' => $this->db->insert_id(),
            'match_code' => $match_code,
            'questions' => $questions
        ];
    }

    /**
     * Generate unique match code
     */
    private function generate_match_code()
    {
        do {
            $code = strtoupper(substr(md5(uniqid(mt_rand(), true)), 0, 6));
            $exists = $this->db->where('match_code', $code)->count_all_results('tbl_battle_1v1');
        } while ($exists > 0);
        
        return $code;
    }

    /**
     * Get random questions for battle
     */
    private function get_random_questions($category_id, $count)
    {
        $questions = $this->db
            ->select('id, question, optiona, optionb, optionc, optiond, answer, note')
            ->where('category', $category_id)
            ->order_by('RAND()')
            ->limit($count)
            ->get('tbl_question')
            ->result_array();
        
        return $questions;
    }

    /**
     * Join a 1v1 battle
     */
    public function join_1v1_battle($match_code, $opponent_id)
    {
        $battle = $this->db
            ->where('match_code', $match_code)
            ->where('status', 'waiting')
            ->where('expires_at >', date('Y-m-d H:i:s'))
            ->get('tbl_battle_1v1')
            ->row_array();
        
        if (!$battle) {
            return ['error' => true, 'message' => 'Battle not found or expired'];
        }
        
        if ($battle['challenger_id'] == $opponent_id) {
            return ['error' => true, 'message' => 'Cannot join your own battle'];
        }
        
        $this->db
            ->where('id', $battle['id'])
            ->update('tbl_battle_1v1', [
                'opponent_id' => $opponent_id,
                'status' => 'ready'
            ]);
        
        $battle['opponent_id'] = $opponent_id;
        $battle['status'] = 'ready';
        $battle['questions'] = json_decode($battle['questions'], true);
        
        return ['error' => false, 'battle' => $battle];
    }

    /**
     * Start 1v1 battle
     */
    public function start_1v1_battle($battle_id)
    {
        return $this->db
            ->where('id', $battle_id)
            ->update('tbl_battle_1v1', [
                'status' => 'playing',
                'started_at' => date('Y-m-d H:i:s')
            ]);
    }

    /**
     * Submit 1v1 battle answers
     */
    public function submit_1v1_answers($battle_id, $user_id, $answers, $score, $correct, $time_ms)
    {
        $battle = $this->get_1v1_battle($battle_id);
        if (!$battle) {
            return ['error' => true, 'message' => 'Battle not found'];
        }
        
        $is_challenger = $battle['challenger_id'] == $user_id;
        $is_opponent = $battle['opponent_id'] == $user_id;
        
        if (!$is_challenger && !$is_opponent) {
            return ['error' => true, 'message' => 'Not a participant'];
        }
        
        $update = [];
        if ($is_challenger) {
            $update['challenger_answers'] = json_encode($answers);
            $update['challenger_score'] = $score;
            $update['challenger_correct'] = $correct;
            $update['challenger_time_ms'] = $time_ms;
            $update['challenger_ready'] = 1;
        } else {
            $update['opponent_answers'] = json_encode($answers);
            $update['opponent_score'] = $score;
            $update['opponent_correct'] = $correct;
            $update['opponent_time_ms'] = $time_ms;
            $update['opponent_ready'] = 1;
        }
        
        $this->db->where('id', $battle_id)->update('tbl_battle_1v1', $update);
        
        // Check if both players finished
        $battle = $this->get_1v1_battle($battle_id);
        if ($battle['challenger_ready'] && $battle['opponent_ready']) {
            $this->complete_1v1_battle($battle_id);
        }
        
        return ['error' => false, 'message' => 'Answers submitted'];
    }

    /**
     * Complete 1v1 battle and determine winner
     */
    private function complete_1v1_battle($battle_id)
    {
        $battle = $this->get_1v1_battle($battle_id);
        
        $winner_id = null;
        $is_draw = 0;
        
        if ($battle['challenger_score'] > $battle['opponent_score']) {
            $winner_id = $battle['challenger_id'];
        } else if ($battle['opponent_score'] > $battle['challenger_score']) {
            $winner_id = $battle['opponent_id'];
        } else {
            // Same score - check time
            if ($battle['challenger_time_ms'] < $battle['opponent_time_ms']) {
                $winner_id = $battle['challenger_id'];
            } else if ($battle['opponent_time_ms'] < $battle['challenger_time_ms']) {
                $winner_id = $battle['opponent_id'];
            } else {
                $is_draw = 1;
            }
        }
        
        $this->db
            ->where('id', $battle_id)
            ->update('tbl_battle_1v1', [
                'winner_id' => $winner_id,
                'is_draw' => $is_draw,
                'status' => 'completed',
                'ended_at' => date('Y-m-d H:i:s')
            ]);
        
        // Award coins to winner
        if ($winner_id && $battle['prize_coins'] > 0) {
            $this->db->query("UPDATE tbl_users SET coins = coins + {$battle['prize_coins']} WHERE id = $winner_id");
        }
        
        return $winner_id;
    }

    /**
     * Get 1v1 battle by ID
     */
    public function get_1v1_battle($battle_id)
    {
        return $this->db
            ->where('id', $battle_id)
            ->get('tbl_battle_1v1')
            ->row_array();
    }

    /**
     * Get 1v1 battle by match code
     */
    public function get_1v1_by_code($match_code)
    {
        return $this->db
            ->where('match_code', $match_code)
            ->get('tbl_battle_1v1')
            ->row_array();
    }

    /**
     * Get user's 1v1 battle history
     */
    public function get_user_1v1_history($user_id, $limit = 20)
    {
        return $this->db
            ->where('challenger_id', $user_id)
            ->or_where('opponent_id', $user_id)
            ->where('status', 'completed')
            ->order_by('ended_at', 'DESC')
            ->limit($limit)
            ->get('tbl_battle_1v1')
            ->result_array();
    }

    // ============================================
    // GROUP BATTLES
    // ============================================

    /**
     * Create a group battle
     */
    public function create_group_battle($group_id, $created_by, $topic_id, $category_id, $config = [])
    {
        $data = [
            'group_id' => $group_id,
            'topic_id' => $topic_id,
            'category_id' => $category_id,
            'created_by' => $created_by,
            'title' => $config['title'] ?? null,
            'question_count' => $config['question_count'] ?? 10,
            'time_per_question' => $config['time_per_question'] ?? 15,
            'entry_coins' => $config['entry_coins'] ?? 0,
            'prize_coins' => $config['prize_coins'] ?? 0,
            'min_players' => $config['min_players'] ?? 2,
            'max_players' => $config['max_players'] ?? 10,
            'status' => 'pending',
            'scheduled_start' => $config['scheduled_start'] ?? null
        ];
        
        // Get random questions
        $questions = $this->get_random_questions($category_id, $data['question_count']);
        $data['questions'] = json_encode($questions);
        
        $this->db->insert('tbl_group_battle', $data);
        $battle_id = $this->db->insert_id();
        
        // Auto-join creator
        $this->join_group_battle($battle_id, $created_by);
        
        return $battle_id;
    }

    /**
     * Join a group battle
     */
    public function join_group_battle($battle_id, $user_id)
    {
        // Check if already joined
        $existing = $this->db
            ->where('battle_id', $battle_id)
            ->where('user_id', $user_id)
            ->get('tbl_group_battle_entry')
            ->row_array();
        
        if ($existing) {
            return false;
        }
        
        $this->db->insert('tbl_group_battle_entry', [
            'battle_id' => $battle_id,
            'user_id' => $user_id,
            'status' => 'joined'
        ]);
        
        // Update player count
        $count = $this->db
            ->where('battle_id', $battle_id)
            ->count_all_results('tbl_group_battle_entry');
        
        $this->db
            ->where('id', $battle_id)
            ->update('tbl_group_battle', ['player_count' => $count]);
        
        return true;
    }

    /**
     * Start a group battle
     */
    public function start_group_battle($battle_id)
    {
        $battle = $this->get_group_battle($battle_id);
        
        if ($battle['player_count'] < $battle['min_players']) {
            return ['error' => true, 'message' => 'Not enough players'];
        }
        
        // Update all entries to playing
        $this->db
            ->where('battle_id', $battle_id)
            ->update('tbl_group_battle_entry', ['status' => 'playing']);
        
        $this->db
            ->where('id', $battle_id)
            ->update('tbl_group_battle', [
                'status' => 'active',
                'started_at' => date('Y-m-d H:i:s')
            ]);
        
        return ['error' => false, 'message' => 'Battle started'];
    }

    /**
     * Submit group battle answers
     */
    public function submit_group_battle_answers($battle_id, $user_id, $answers, $score, $correct, $wrong, $time_ms)
    {
        $this->db
            ->where('battle_id', $battle_id)
            ->where('user_id', $user_id)
            ->update('tbl_group_battle_entry', [
                'answers' => json_encode($answers),
                'score' => $score,
                'correct_answers' => $correct,
                'wrong_answers' => $wrong,
                'total_time_ms' => $time_ms,
                'status' => 'completed',
                'completed_at' => date('Y-m-d H:i:s')
            ]);
        
        // Check if all players finished
        $pending = $this->db
            ->where('battle_id', $battle_id)
            ->where('status !=', 'completed')
            ->count_all_results('tbl_group_battle_entry');
        
        if ($pending == 0) {
            $this->complete_group_battle($battle_id);
        }
        
        return true;
    }

    /**
     * Complete group battle and calculate rankings
     */
    private function complete_group_battle($battle_id)
    {
        // Get all entries sorted by score (desc) and time (asc)
        $entries = $this->db
            ->where('battle_id', $battle_id)
            ->order_by('score', 'DESC')
            ->order_by('total_time_ms', 'ASC')
            ->get('tbl_group_battle_entry')
            ->result_array();
        
        $battle = $this->get_group_battle($battle_id);
        
        // Assign ranks and coins
        $rank = 1;
        foreach ($entries as $entry) {
            $coins = 0;
            if ($rank == 1 && $battle['prize_coins'] > 0) {
                $coins = $battle['prize_coins'];
                $this->db->query("UPDATE tbl_users SET coins = coins + $coins WHERE id = {$entry['user_id']}");
            }
            
            $this->db
                ->where('id', $entry['id'])
                ->update('tbl_group_battle_entry', [
                    'rank' => $rank,
                    'coins_earned' => $coins
                ]);
            
            $rank++;
        }
        
        // Mark battle as completed
        $this->db
            ->where('id', $battle_id)
            ->update('tbl_group_battle', [
                'status' => 'completed',
                'ended_at' => date('Y-m-d H:i:s')
            ]);
    }

    /**
     * Get group battle by ID
     */
    public function get_group_battle($battle_id)
    {
        return $this->db
            ->where('id', $battle_id)
            ->get('tbl_group_battle')
            ->row_array();
    }

    /**
     * Get group battle entries
     */
    public function get_group_battle_entries($battle_id)
    {
        return $this->db
            ->select('e.*, u.name, u.profile')
            ->from('tbl_group_battle_entry e')
            ->join('tbl_users u', 'u.id = e.user_id')
            ->where('e.battle_id', $battle_id)
            ->order_by('e.rank', 'ASC')
            ->get()
            ->result_array();
    }

    /**
     * Get group's battle history
     */
    public function get_group_battles($group_id, $status = null, $limit = 20)
    {
        $this->db->where('group_id', $group_id);
        
        if ($status) {
            $this->db->where('status', $status);
        }
        
        return $this->db
            ->order_by('created_at', 'DESC')
            ->limit($limit)
            ->get('tbl_group_battle')
            ->result_array();
    }

    /**
     * Get pending battles for a group
     */
    public function get_pending_battles($group_id)
    {
        return $this->db
            ->where('group_id', $group_id)
            ->where('status', 'pending')
            ->order_by('scheduled_start', 'ASC')
            ->get('tbl_group_battle')
            ->result_array();
    }
}

